rasa/utils/tensorflow/layers.py

```diff
@@ -949,3 +949,232 @@ def call(
         )
 
         return loss, accuracy
+
+
+class MultiLabelDotProductLoss(DotProductLoss):
+    """Same as DotProductLoss but adapted for multiple labels for a single data point."""
+
+    def _get_neg_values(self, tensor, neg_ids):
+
+        tensor_expanded = tf.tile(
+            tf.expand_dims(self._make_flat(tensor), 0), (tf.shape(tensor)[0], 1, 1)
+        )
+        neg_tensor = tf.gather(tensor_expanded, neg_ids, batch_dims=1)
+        neg_tensor = tf.expand_dims(neg_tensor, axis=1)
+
+        return neg_tensor
+
+    def _sample_negatives(
+        self,
+        inputs_embed: tf.Tensor,
+        labels_embed: tf.Tensor,
+        label_ids: tf.Tensor,
+        all_labels_embed: tf.Tensor,
+        all_label_ids: tf.Tensor,
+    ) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor, tf.Tensor]:
+        """Sample negative examples."""
+
+        pos_inputs_embed = tf.expand_dims(
+            inputs_embed, axis=-2, name="expand_pos_input"
+        )
+        # Take only the first label as positive, others may or may not be picked during random sampling
```

Comment:
why only the first one?

Reply:
The others(some of them) will be picked up during negative sampling.
So, the resultant `pos_neg_labels`(at the end of the function) will contain some 1s and some 0s depending on which data points were sampled during negative sampling.

Reply:
but negative sampling is independent, no? We just later create a mask that would set loss from same labels to 0, or do I miss anything?

Reply:
We don't set the loss to 0, but actually set the correct label for it.

So, initially if the label_ids passed to this function were - 
```
[[2,3,0,0],
[1,0,0,0],
[3,1,2,0]]
```
(Here `0` is the padding label ID)

and if `neg_ids` are(`num_negs=2`) - 
```
[[1,2],
[0,1],
[2,0]]
```

then `neg_labels_ids` would be - 
```
[[1,1],
[2,1],
[1,2]]
```

and `pos_neg_labels` will become - 
```
[[0,0],
[0,1],
[1,1]]
```

Now, the final `label_ids` that are constructed [here](https://github.com/RasaHQ/rasa/pull/7399/files#diff-dd1545c3aa7b4edad5ad9b0e3f9090a5413e0309af2831e0bfe25b50f644888eR1057) and get passed to `sigmoid_cross_entropy_with_logits` are - 
```
[[1,0,0],
[1,0,1],
[1,1,1]]
```

As you can see the final `label_ids` do not have the same number of positive labels as they were present in the data but the model will definitely see all positive labels in some other batch, due to random sampling. So, the sampling that we do here is not strictly "negative". Does that help clarify?

Reply:
hm, I need to spend some time on it 🙈